static/js/util.ts

```diff
@@ -39,54 +44,69 @@ export function lower_bound(array, value, less) {
     return first;
 }
 
-export const lower_same = function lower_same(a, b) {
+export const lower_same = function lower_same(a?: string, b?: string): boolean {
     if (a === undefined || b === undefined) {
-        blueslip.error(`Cannot compare strings; at least one value is undefined: ${a}, ${b}`);
+        blueslip.error(
+            `Cannot compare strings; at least one value is undefined: \
+${a ?? "(undefined)"}, ${b ?? "(undefined)"}`,
+        );
         return false;
     }
     return a.toLowerCase() === b.toLowerCase();
 };
 
-export const same_stream_and_topic = function util_same_stream_and_topic(a, b) {
+export type StreamTopic = {
+    stream_id: number;
+    topic: string;
+};
+
+export const same_stream_and_topic = function util_same_stream_and_topic(
+    a: StreamTopic,
+    b: StreamTopic,
+): boolean {
     // Streams and topics are case-insensitive.
     return a.stream_id === b.stream_id && lower_same(a.topic, b.topic);
 };
 
-export function is_pm_recipient(user_id, message) {
-    const recipients = message.to_user_ids.split(",");
+export function is_pm_recipient(user_id: number, message_recipient: PMRecipient): boolean {
+    const recipients = message_recipient.to_user_ids.split(",");
     return recipients.includes(user_id.toString());
 }
 
-export function extract_pm_recipients(recipients) {
+export function extract_pm_recipients(recipients: string): string[] {
     return recipients.split(/\s*[,;]\s*/).filter((recipient) => recipient.trim() !== "");
 }
 
-export const same_recipient = function util_same_recipient(a, b) {
+// When the type is "private", properties from PMRecipient, namely to_user_ids might be
+// undefined.
+type Recipient = (Partial<PMRecipient> & {type: "private"}) | (StreamTopic & {type: "stream"});
```

Comment:
`Partial<PMRecipient>` is an odd thing to do. It seems like we may have found an actual bug here: according to this comment, `to_user_ids` can be left `undefined` for semi-legitimate reasons.

https://github.com/zulip/zulip/blob/5e74a8d0cc8f78f950e71a93a7c5b034d8a805a0/static/js/compose.js#L156-L162

So `util.is_pm_recipient` probably needs to handle that case? If that’s right, we should declare `to_user_ids` an optional property of `PMRecipient` and remove the `Partial<>`.

Reply:
Hmm, I missed this case because I only tracked down the callers of `util.is_pm_recipient`, but it does not need to be handled. The only place where it is called is `compose_fade_helper.js`, and `focused_recipient` is always populated with `to_user_ids`. We can move `is_pm_recipient` to `compose_fade_helper` as a prep commit, but the type change to `PMRecipient` still applies.

Reply:
That change works for me. One small GitHub style comment: It'd have been better to clarify that you in fact did this tweak in a new commit; it took me a while to realize this proposal was already implemented.